text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,602 @@
+differences to `return`, and give examples of possible use cases and mistakes.
+
+
+## Discuss how this impacts the ability to read, understand, and maintain Rust code. Code is read and modified far more often than written; will the proposed feature make code easier to maintain?
+As the `become` keyword is already reserved, this has no impact on existing code. For code
+that does use this feature, it is required that a programmer understands the differences between `become` and `return`,
+it is difficult to judge how big this impact is without an initial implementation. One difference, however, is in
+debugging code that uses `become`. As the stack is not preserved, debugging context is lost which likely makes debugging
+more difficult. That is, elided parent functions as well as their variable values are not available during debugging.
+(Though this issue might be lessened by providing a flag to opt out of TCO, which would, however, break the semantic
+guarantee of not creating stack frames. This is likely an issue that needs some investigation after creating an initial
+implementation.)
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+<!-- This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work. -->
+This explanation is mostly based on a [previous RFC](https://github.com/DemiMarie/rfcs/blob/become/0000-proper-tail-calls.md#detailed-design)
+though is more restricted as the current RFC does not target general tail calls anymore.
+
+The goal of this RFC is to describe a first implementation that is already useful while providing a basis to explore
+possible ways to relax the requirements when TCO can be guaranteed.
+
+## Syntax
+[syntax]: #syntax
+
+A guaranteed TCO is indicated by using the `become` keyword in place of `return`. The `become` keyword is already
+reserved, so there is no backwards-compatibility break. The `become` keyword must be followed by a plain function call
+or method calls, that is supported are calls like: `become foo()`, `become foo(a)`, `become foo(a, b)`, and so on, or
+`become foo.bar()` with plain arguments. Neither the function call nor any arguments can be part of a larger expression
+such as `become foo() + 1`, `become foo(1 + 1)`, `become foo(bar())`. Additionally, there is a further restriction on
+the tail-callable functions: the function signature must exactly match that of the calling function. 
+
+Invocations of overloaded operators with at least one non-primitive argument were considered as valid targets, but were
+rejected on grounds of being too error-prone. In any case, these can still be called as methods.
```

Comment:
The problem is that using operators it is not always clear what all the arguments and their types are.

Are you asking for an update on the text or just out of curiousity?

Reply:
I don't think I understand this, how are operators different from function (in traits for example)? It seems like the argument types should be as clear as with any other function...

> Are you asking for an update on the text or just out of curiousity?

Well, I want to understand the concern first. And maybe lift this restriction, if it turns out to be unnecessary :)

Reply:
Thank you for taking the time to question these points!

In my (and [others](https://github.com/rust-lang/rfcs/pull/3407/files/3e7c3847bb092ef365e1d93969a4c7f5d723fdff#r1159977404)) opinion, operators can get confusing.

So for example `SubAssign`:

Operator form would be this:
```rust
become a -= b;
```

Function / method form would be this:
```rust
use std::ops::SubAssign;
// ...

become a.sub_assign(b);
// and
become sub_assign(&mut a, b);
```

For me the first version looks like too much magic, while either of the second versions seem quite clear.

So while the restriction is not strictly necessary, I think it improves readability and maintainability.

Reply:
I don't *really* see the difference between, say `return a + b` and `become a + b`, you could say that `become` only works with functions and `+` is not a function... but in the end of the day it *is* a function, just with a weird syntax.

Reply:
Ideal, `become` works with (somehow) recursive functions only:
```rust
fn recursive_fn1(a : i32) { /* ... */ return recursive_fn1(a) + 5;}
// impossible to do TCE

fn recursive_fn2(a : i32) { /* ... */ return recursive_fn2(a + 5);} 
// possible to do TCE
```

Reply:
We are agreeing on everything except:
> just with a weird syntax

To me it seems to not be just some weird syntax, I can see this causing confusion. The question is does allowing operators provide a higher benefit than the possible confusion they can cause. Is there a default approach that is normally taken for these kind of decisions?

Reply:
I think it's fine to specify that *overridable* operators in rustc can be `become`'d, since they are analogous to method calls. 

The one point of clarifications is that operators involving built-in numeric types get lowered into [built-in mir operations](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/syntax/enum.Rvalue.html#variant.BinaryOp), so it's not guaranteed that `become 1 + 2`, for example, will emit a `jmp` (not sure if that's guaranteed in general for `become`, anyways, though).

Reply:
> I can see this causing confusion

Sorry, I still can't understand what confusion can this cause...

Reply:
Here is an example that could be confusing if we allowed overloadable operators to be used with `become`:
```rust
pub fn fibonacci(n: u64) -> u64 {
    if n <= 2 {
        return n
    }
    become fibonacci(n - 2) + fibonacci(n - 1)
}
```
In this case a naive author might assume that this is going to be a stack space efficient implementation since it uses tail recursion instead of normal recursion. However, the outcome is more of less the same since the critical recursive calls are not actually in tail call position.

Reply:
@Robbepop thanks for the example! This answers my question & looks convincing enough, supporting operators is not important (given that you can always replace them with method calls), while avoiding such confusion and providing diagnostics here might be valuable. <sub>Resolving this conversation</sub>

Reply:
@WaffleLapkin Also, operators don't always desugar to method calls. `1 + 2` is not a method call. This is even observable for `+=` because the primitives have a different order of evaluation than the method call:
```rust
let mut x = 0u32;
*dbg!(&mut x) += dbg!(1); // 1 prints, then &mut x
let mut x = String::from("hi");
*dbg!(&mut x) += dbg!("fi"); // &mut x prints, then "fi"
```

Reply:
@Robbepop thanks for your example! I added it to the RFC.